tsnake/snake.py

```diff
@@ -52,59 +80,147 @@ def intersects_grid_cell(self, grid_cell):
 class TSnake(object):
     """
     A class representing a *single* T-snake at a given instance in time.
-    
+
     If the initial T-snake splits into two, then there should be *two*
       instances of T-snakes present in the algorithm.
       (NOTE: can change this if inconvenient)
-      
+
     The merging two T-snakes in the algorithm should be done with:
       TSnake.merge(snake1, snake2)
-      
+
     In this class, 
         each element/edge       self.elements[i] 
             corresponds to:
         node pair               (self.nodes[i], self.nodes[i+1])
-    
+
     Args:
     ===========================================
     (list) nodes:
     * A list of Node instances, in order i=0, 1, 2, ..., N-1
     ===========================================
-    
+
     TODO: hyperparameters in constructor
     TODO: calculate intensity normal thingy
     TODO: comments for a,b, gamma
     """
-    
+
     def __init__(self, nodes, force, intensity, a, b, gamma, dt):
         for n in nodes:
             assert isinstance(n, Node)
         self.nodes = list(nodes)
         # Force and intensity fields over the image, (n,m) np arrays
         self.force = force
         self.intensity = intensity
-        #Deformation parameters
+        # Deformation parameters
         self.a = a
         self.b = b
         self.gamma = gamma
         self.dt = dt
-        
+
         self._elements = []
+        
         # Connect each node[i] --> node[i+1]
         for i in range(len(nodes)-1):
             self._elements.append(Element(self.nodes[i], self.nodes[i+1]))
-            
+
         # Connect node[N-1] --> node[0]
         self._elements.append(Element(self.nodes[-1], self.nodes[0]))
+        self._compute_normals()
+
+    def _compute_normals(self):
+        """
+        Compute normals
+        TODO: Finish doc, see if we can optimize this,
+        Computation is O(n) in the number of edges
+        Returns np array of length(# nodes, 2) representing the normal at each node
+        """
+
+        for i in range(len(self._elements)):
```

Comment:
I'm kind of confused about what is happening here. We are calculating the normals of the snake, which should be a unit vector that is perpendicular to the snake at a given point and points to the outside of the snake. I see you are checking for intersections between elements. Could you give me a high-level description of what the algorithm is doing?

Reply:
It's actually not a non-trivial problem, and a full explanation would be pretty verbose. Put simply: 

1. Figure out the direction of our first normal. We can do this by guessing the direction, and then checking where this line interesects with the snake.
2. If the direction of our intersection is the same as the direction of our normal, we must be pointing inside, so flip.
3. Else we are not, continue
4. For each subsequent normal, use similar intersection logic but for normals instead of snake elements. 

You can visualize the result if you run the jupyter notebook I added.

Reply:
It might also help if you think about how you would solve the problem to understand what's going on.